### ISSUE TITLE
Detail shortcomings of `zfs diff` in order to mitigate

### ISSUE THREAD

**Issue by kimono-koans** [openzfs/zfs#14990]
### System information

Type | Version/Name
 --- | ---
Distribution Name	|   Ubuntu
Distribution Version	|  22.04
Kernel Version	|  5.19.0
Architecture	|  amd64
OpenZFS Version	| zfs-2.1.5-1ubuntu6~22.04.1, zfs-kmod-2.1.5-1ubuntu6

### Describe the problem you're observing

`zfs diff` occasionally fails to show a file creation or removal, and does not report an error.  Perhaps this is some odd hard link behavior, but hard link search of a snap dataset show relevant files don't seem to be hard links?

### Describe how to reproduce the problem

`sudo httm --roll-forward=snap_2023-06-08-18:23:09_ProgrammingSync` attempts to copy the files designated as changed by `zfs diff` and preserve hard links by a method similar to `rsync`.  After parsing `zfs diff` and copying/removing/linking files:

```
➜  httm git:(master) ✗ rsync -rincv /srv/program/.zfs/snapshot/snap_2023-06-08-18:23:09_ProgrammingSync/ /srv/program
sending incremental file list
skipping non-regular file "Rust Book Practice/hello/target/rls/debug/build/libc-1521b2b94739c351/build-script-build.dSYM"
skipping non-regular file "Rust Book Practice/hello/target/rls/debug/build/libc-f58bd583ca515ff9/build-script-build.dSYM"
skipping non-regular file "Rust Book Practice/hello/target/rls/debug/build/signal-hook-64d2e5ac31d5c081/build-script-build.dSYM"
skipping non-regular file "Rust Book Practice/hello/target/rls/debug/build/signal-hook-6a1c4b257beb5b7c/build-script-build.dSYM"
>fcsT...... dano/target/release/deps/dano-1ebe2f3c696ea7d0
>f+++++++++ httm/target/release/deps/httm-2171e5bd81d234f5
>f+++++++++ httm/target/release/deps/httm-3ee48a9d66b3242e
>f+++++++++ httm/target/release/deps/httm-4fd6be4cbf63384c
>f+++++++++ httm/target/release/deps/httm-5973f3bd92b6240a
>f+++++++++ httm/target/release/deps/httm-7d01d37c9534ae62
>f+++++++++ httm/target/release/deps/httm-7d328ace00d67b8b
>f+++++++++ httm/target/release/deps/httm-8631aafb71fc7bbe
>f+++++++++ httm/target/release/deps/httm-92bce85749d5e2dd
>f+++++++++ httm/target/release/deps/httm-981bedbb8ad8c838
>f+++++++++ httm/target/release/deps/httm-9a8972972957ad07
>f+++++++++ httm/target/release/deps/httm-a7983a6a5cddf68c
>f+++++++++ httm/target/release/deps/httm-b248f0bb38df2a84
>f+++++++++ httm/target/release/deps/httm-b7b4cd3d573e048d
>f+++++++++ httm/target/release/deps/httm-c4aca72a97722449
>f+++++++++ httm/target/release/deps/httm-c5d662acf83f1272
>f+++++++++ httm/target/release/deps/httm-cb5a43548b3e30ad
>f+++++++++ httm/target/release/deps/httm-cc5565dc4e425e6d
>f+++++++++ httm/target/release/deps/httm-e1f713d5a6acda31
>f+++++++++ httm/target/release/deps/httm-f93b961633f04d88

sent 1,805,898 bytes  received 6,441 bytes  43,670.82 bytes/sec
total size is 14,765,608,685  speedup is 8,147.27 (DRY RUN)

➜  httm git:(master) ✗ zfs diff -H -t -h rpool/program@snap_2023-06-08-18:23:09_ProgrammingSync 2&>1 | grep httm-3ee48a9d66b3242e
1685076912.894943601	-	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e/invoked.timestamp
1685076913.358942412	-	/srv/program/httm/target/release/deps/httm-3ee48a9d66b3242e.d
1685077029.998644933	-	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e
1685077029.998644933	-	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e/dep-bin-httm
1685077029.998644933	-	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e/bin-httm
1685077029.998644933	-	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e/bin-httm.json
1687008800.080815244	+	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e
1687008792.556836496	+	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e/bin-httm
1687008792.720836032	+	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e/bin-httm.json
1687008793.932832608	+	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e/dep-bin-httm
1687008794.888829907	+	/srv/program/httm/target/release/.fingerprint/httm-3ee48a9d66b3242e/invoked.timestamp
1687008801.880810161	+	/srv/program/httm/target/release/deps/httm-3ee48a9d66b3242e.d
➜  httm git:(master) ✗ cd /srv/program/.zfs/snapshot/snap_2023-06-08-18:23:09_ProgrammingSync/
➜  snap_2023-06-08-18:23:09_ProgrammingSync find . -type f -links +1 -printf 'inode %i links %n name %p\n' | grep httm-3ee48a9d66b3242e
➜  snap_2023-06-08-18:23:09_ProgrammingSync
```

**Comment by rincebrain** [comment]
You may find [this talk](https://openzfs.org/wiki/OpenZFS_Developer_Summit_2020_talks#Improved_.E2.80.9Czfs_diff.E2.80.9D_performance_with_reverse-name_lookup_.28Sanjeev_Bagewadi_.26_David_Chen.29) informative.

**Comment by scineram** [comment]
> You may find [this talk](https://openzfs.org/wiki/OpenZFS_Developer_Summit_2020_talks#Improved_.E2.80.9Czfs_diff.E2.80.9D_performance_with_reverse-name_lookup_.28Sanjeev_Bagewadi_.26_David_Chen.29) informative.

Would recommend #12837 too!

**Comment by kimono-koans** [comment]
Just to put a button in this, as it may not be obvious to others, what are the shortcomings I found re: the `zfs diff` model.  

These are the `zfs diff` shortcomings I found while [implementing `httm --roll-forward`](src/exec/roll_forward.rs):

1. As noted in the linked talk, one needs to do hard link discovery on the snapshot and live datasets to discover hard links and re-link/un-link them.

2. However -- less obvious -- one needs to keep track of potentially orphaned links (all files with one unique inode), as how `zfs diff` will do path resolution will sometimes not indicate the linked path has been removed, etc.

Thanks.  Appreciate your help.
